### PR TITLE
[jit] refactor register_prim_ops

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21003 [jit] remove constant pooling expect
* #21002 [jit] move casting ops from prim to aten
* **#21001 [jit] refactor register_prim_ops**

Differential Revision: [D15523445](https://our.internmc.facebook.com/intern/diff/D15523445)

### CODE REVIEW

**Review by driazati — APPROVED** [review]
Seems fine, though it's all temporary until we land something along the lines of #19373, bind all these as custom ops, and delete all these macros.

#19373 was blocked on removing assumptions that `prim::` and `aten::` ops are pure unless annotated otherwise (they should match custom ops, which are assumed to be side-effectful unless specified)

**Comment by suo** [comment]
@driazati the purity thing is fixed btw so that PR is no longer blocked

**Pull request closed without merging**
